opentelemetry-sdk/src/opentelemetry/sdk/_metrics/export/__init__.py

```diff
@@ -96,6 +96,36 @@ def shutdown(self) -> None:
         pass
 
 
+class InMemoryMetricReader(MetricReader):
+    """Implementation of :class:`MetricReader` that returns its metrics from :func:`metrics`.
+
+    This is useful for e.g. unit tests.
+    """
+
+    def __init__(
+        self,
+        preferred_temporality: AggregationTemporality = AggregationTemporality.CUMULATIVE,
+    ) -> None:
+        super().__init__(preferred_temporality=preferred_temporality)
+        self._lock = RLock()
+        self._metrics: List[Metric] = []
```

Comment:
Similarly to the inmemoryexporter for spans, shouldn't the exporter be the one that contains the metrics in memory? I can see a use case for a `SimpleMetricReader` in which we want to collect on demand, which `collect()` would be called manually. So this can be used in conjuction with a `InmemoryMetricsExporter` which has a sequence of the metrics in memory which can be used for testing. Thoughts?

Reply:
Yes I agree with pull exporter/metric reader being a better fit for unit testing. 

> Is there a use case you can think of for a push based InMemoryMetricsExporter?

Not specifically the InMemoryMetricsExporter. I know there are several azure customers who want a push based `SimpleMetricReader ` as described above. Just thought it would be a good time to implement this in conjunction with the `InmemoryMetricsExporter` if we decided to use a push-based `InMemoryMetricsExporter` for testing. Not a blocker though.

Reply:
Didn't 100% follow you on `SimpleMetricReader`. Would it be equivalent to a PeriodicExportingMetricReader with an export interval of infinity?

Reply:
I think the scenario would be:
```
exporter = CustomMetricsExporter()
reader = SimpleMetricReader(exporter)
provider = MeterProvider(metric_readers=[reader])
meter = get_meter_provider().get_meter(__name__)

counter = meter.create_counter("counter")
counter.add(1)

reader.collect() -> collects and then exports immediately
```

> Would it be equivalent to a PeriodicExportingMetricReader with an export interval of infinity?

Possibly could achieve the same thing but without the overhead of another thread.

Reply:
I understand that it may be better for the testing/debugging purposes to implement a `MetricReader` instead of a `MetricExporter`. Nevertheless, this is not spec compliant, the spec defines an `InMemoryMetricExporter`.

I am not against merging this, I just don't think we should consider the issue to be closed if this PR gets merged.

Reply:
> Possibly could achieve the same thing but without the overhead of another thread.

@lzchen that makes sense. The only thing missing there is a timeout which you would need a background thread/coroutine for regardless.

> I understand that it may be better for the testing/debugging purposes to implement a MetricReader instead of a MetricExporter

@ocelotl we are using metric reader as pull exporter as permitted in the spec. The spec does say it should be a push exporter which [doesn't make sense to me](https://github.com/open-telemetry/opentelemetry-specification/issues/2425).

Reply:
> @ocelotl we are using metric reader as pull exporter as permitted in the spec. The spec does say it should be a push exporter which [doesn't make sense to me](https://github.com/open-telemetry/opentelemetry-specification/issues/2425).

Sorry, I can't find where in the spec this is permitted, can you point to that part, please?

Again, I am not against merging this, I just think to be spec compliant we will need a push exporter as the spec says.
Just for the record, I would prefer the spec don't even have this requirement at all, but if we are to implement something to help ourselves or our users debug or test, I also think a pull metric reader would be a better choice.

Reply:
Yup, see [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#:~:text=in%20the%20SDK%3A-,Model%20the%20pull%20exporter%20as%20MetricReader,-%2B%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2B%20%20%20%20%20%20%20%20%20%20%20%20%2B%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2D%2B%0A%7C%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7C%20Metrics...%20%7C%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7C%0A%7C%20In%2Dmemory) 

> Model the pull exporter as MetricReader
> 
> ```
> +-----------------+            +-----------------------------+
> |                 | Metrics... |                             |
> | In-memory state +------------> PrometheusExporter (pull)   +---> Another process (scraper)
> |                 |            | (modeled as a MetricReader) |
> +-----------------+            |                             |
>                                +-----------------------------+
> ```

Agreed on just removing the requirement. Do you mind leaving a comment on https://github.com/open-telemetry/opentelemetry-specification/issues/2425?

Reply:
Sure, replied in the spec already :v: